x-pack/plugins/enterprise_search/server/lib/get_oauth_token_package_params.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ENTERPRISE_SEARCH_KIBANA_COOKIE } from '../../common/constants';
+
+export function getOAuthTokenPackageParams(cookieHeaders: string | string[] | undefined) {
+  // In the future the token package will be stored in the login session. For now it's in a cookie.
+
+  // No cookie headers? No tokens
+  if (!cookieHeaders) {
+    return {};
+  }
+
+  // Take any cookie headers and split the individual cookies out, e.g. "_my_cookie=chocolateChip"
+  const cookiePayloads = [cookieHeaders].flat().flatMap((rawHeader) => rawHeader.split('; '));
+
+  // Split those raw cookies into [key, value] pairs, e.g. ["_my_cookie", "chocolateChip"]
+  const cookiePairs = cookiePayloads.map((rawCookie) => rawCookie.split('='));
+
+  const tokenPackageCookie = cookiePairs.find((cookiePair) => {
+    return cookiePair[0] === ENTERPRISE_SEARCH_KIBANA_COOKIE;
+  });
+
+  if (tokenPackageCookie) {
+    return { token_package: tokenPackageCookie[1] };
```

Comment:
This is a lot of separate iteration we're doing (3 different iterations - `flatMap`, `map`, then `find`). It's not 100% clear to me why we're doing so - if we're just targeting a single cookie, which is `ENTERPRISE_SEARCH_KIBANA_COOKIE`, why not just look for it the first time when we're iterating through the cookie array?

In this case I do think we need the `flatMap`, but I think we don't need an extra `find` and could do our searching in the 2nd iteration:

```suggestion
  let tokenPackage;

  cookiePayloads.forEach((rawCookie) => {
    const [key, value] = rawCookie.split('=');
    if (key === ENTERPRISE_SEARCH_KIBANA_COOKIE) tokenPackage = value;
  });

  if (tokenPackage) {
    return { token_package: tokenPackage };
```

Reply:
Sure, I can combine a few of these iterations. I was mostly trying to strike a balance between readability (making it clearer what I was doing to parse the cookie header(s)) and efficiency. If you find this more readable, I'll change to something more like it.

Reply:
Gotcha, I can see how the `.find` is there for readability - in this case I do strongly think `const [key, value] = rawCookie.split('=');` is more readable (for those familiar w/ JS destructuring) & avoids unclear references to `[0]`/`[1]` indexes. Removing an extra iteration is a microperf bonus on top of that.